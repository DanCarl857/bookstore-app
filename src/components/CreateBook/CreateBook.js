import React, { useState } from 'react';
import { graphql } from 'react-apollo';
import { compose } from 'recompose';
import { Button, Form, FormGroup, Label, Input } from 'reactstrap';
import { Link } from 'react-router-dom';
import { createBookMutation } from './../../queries/queries';
import * as UTILS from './../../utils';

const CreateBook = ({ createBookMutation, history }) => {

    const [title, setTitle] = useState('');
    const [author, setAuthor] = useState('');
    const [price, setPrice] = useState('');
    const [errors, setErrors] = useState({
        title: '',
        author: '',
        price: ''
    });

    const submitForm = (event) => {
        event.preventDefault();
        if (UTILS.validateForm(errors)) {
            createBookMutation({
                variables: {
                    title,
                    author,
                    price
                }
            }).then(() => history.push('/'));
        } else {
            console.log('Invalid Form');
        }

    };

    const handleChange = (event) => {
        event.preventDefault();
        const { name, value } = event.target;
        let tempErrors = {...errors };

        switch (name) {
            case 'title':
                tempErrors.title = value.length < 5 ? 'Title should be at least 5 characters long.' : '';
                setTitle(value);
                break;
            case 'author':
                tempErrors.author = value.length < 5 ? 'Author name should be at least 5 characters long.' : '';
                setAuthor(value);
                break;
            case 'price':
                tempErrors.price = parseFloat(value) <= 0 ? 'Book cannot be free.' : '';
                setPrice(price);
                break;
            default:
                break;
        }
        setErrors(tempErrors);
    }

    return (
        <div className="container box">
            <h2>Create a New Book</h2>
            <hr />
            <Form onSubmit={(event) => submitForm(event)}>
                <FormGroup>
                    <Label for="title">Title</Label>
                    <Input
                        type="text"
                        name="title"
                        id="title"
                        placeholder="Enter book title"
                        onChange={(e) => handleChange(e)}
                    />
                    {errors.title.length > 0 && <span className="error">{errors.title}</span>}
                </FormGroup>
                <FormGroup>
                    <Label for="author">Author</Label>
                    <Input
                        type="text"
                        name="author"
                        id="author"
                        placeholder="Enter author's name"
                        onChange={(e) => handleChange(e)}
                    />
                    {errors.author.length > 0 && <span className="error">{errors.author}</span>}
                </FormGroup>
                <FormGroup>
                    <Label for="price">Price</Label>
                    <Input
                        type="number"
                        name="price"
                        id="price"
                        placeholder="Enter book price"
                        onChange={(e) => handleChange(e)}
                    />
                    {errors.price.length > 0 && <span className="error">{errors.price}</span>}
                </FormGroup>
                <Button color="success" size="sm">Create Book</Button>{'  '}
                <Button color="danger" size="sm">
                    <Link to="/">Cancel</Link>
                </Button>
            </Form>
        </div>
    )
}

export default compose(
    graphql(createBookMutation, { name: "createBookMutation"})
)(CreateBook);