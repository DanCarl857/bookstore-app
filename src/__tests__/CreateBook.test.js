import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import ApolloClient from 'apollo-boost';
import { MemoryRouter } from 'react-router';

import CreateBook from './../components/CreateBook/CreateBook';

const client = new ApolloClient({
    uri: 'http://localhost:4567/graphql'
});

describe('Create Book', () => {

    const component = mount(shallow(
        <ApolloProvider client={client}>
            <MemoryRouter>
                <CreateBook />
            </MemoryRouter>
        </ApolloProvider>
    ).get(0));

    it('should render correctly', () => {
        const component = shallow(<CreateBook />);
        expect(component).toMatchSnapshot();
    });

    it('should have 2 text fields for title and author', () => {
        expect(component.find('input[type="text"]').length).toEqual(2);
    });

    it('should have 1 input field for the price of a book', () => {
        expect(component.find('input[type="number"]').length).toEqual(1);
    });
});