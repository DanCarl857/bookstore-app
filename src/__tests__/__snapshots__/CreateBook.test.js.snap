// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Book should render correctly 1`] = `ShallowWrapper {}`;
